feat(transformer): transform explicit resource management

camc314 opened the pull request:
~~wanted to have a shot at implementing this, this is rough draft.~~

~~Dunqing or overlookmotel feel free to take over or close out this PR. not sure how much time i'll have to finish it off~~

@Dunqing @overlookmotel this should be ready when you guys have time 🙏 

closes #9168

camc314 commented:
* **#9310** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9310?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9310?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9310).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 commented on tasks/transform_conformance/snapshots/babel.snap.md:
```diff
@@ -2744,3 +2744,10 @@ transform-react-jsx: unknown field `autoImport`, expected one of `runtime`, `dev
 x Output mismatch
 
 
+# babel-plugin-proposal-explicit-resource-management (23/24)
+* transform-top-level/hoisting-default-class/input.mjs
+Symbol span mismatch for "C":
+after transform: SymbolId(1): Span { start: 37, end: 38 }
+rebuilt        : SymbolId(2): Span { start: 0, end: 0 }
```
I don't understand why rebuilt has a span of `0,0` 🤷

Dunqing commented:
Thank you for working on this, I will review this next week, before that, can you add some documentation like other plugins does at the top of file?

For example:
https://github.com/oxc-project/oxc/blob/3c1d7f6752a1bb99fd13d80a71b12942404942ee/crates/oxc_transformer/src/es2020/nullish_coalescing_operator.rs#L1-L30

overlookmotel commented on crates/oxc_transformer/src/proposals/explicit_resource_management.rs:
```diff
@@ -0,0 +1,768 @@
+            let old_body = ctx.ast.move_statement(&mut node.body);
+
+            let new_body = ctx.ast.vec_from_array([stmt, old_body]);
+            node.body = Statement::BlockStatement(ctx.ast.alloc_block_statement_with_scope_id(
```
```suggestion
            node.body = ctx.ast.statement_block_statement_with_scope_id(
```

(but rebase on latest main first - Dunqing only added these shortcut APIs earlier this week)

camc314 replied:
have a link to the code ref? i can't see this function

camc314 replied:
nvm found it - `statement_block_with_scope_id`

overlookmotel commented on crates/oxc_transformer/src/proposals/explicit_resource_management.rs:
```diff
@@ -0,0 +1,768 @@
+            node.body = Statement::BlockStatement(ctx.ast.alloc_block_statement_with_scope_id(
+                SPAN,
+                new_body,
+                ctx.create_child_scope(ctx.current_scope_id(), ScopeFlags::empty()),
```
I don't think this is quite correct. `current_scope_id` isn't updated until *after* `enter_for_of_statement` is called:

https://github.com/oxc-project/oxc/blob/4ca62abe43e4b19fa9c9605a4c866242031626fb/crates/oxc_traverse/src/generated/walk.rs#L1802-L1813

So the correct parent scope is `for_of_stmt_scope_id`, not `ctx.current_scope_id()`.


This is rather unintutive, I know! If I remember right, the reason was that for some types the scope is entered late/exited early (scope doesn't start before the 1st field e.g. `Class`). So we chose to always call `enter_*` before entering the scope, and `exit_*` after exiting the scope, so you don't have to remember which types have this "late entry" / "early exit" behavior, and which don't.

If making that change doesn't fix/fail any tests, I guess there's no tests for the case where `for` statement doesn't already have a statement block. If so, we should add one.

camc314 replied:
this is a very good catch! babel had no tests for this so i've added one

overlookmotel commented on crates/oxc_transformer/src/proposals/explicit_resource_management.rs:
```diff
@@ -0,0 +1,768 @@
+
+use std::mem;
+
+use oxc_allocator::Address;
+use oxc_allocator::GetAddress;
+use oxc_ast::ast::BlockStatement;
+use oxc_ast::ast::Class;
+use oxc_ast::ast::ClassType;
+use oxc_ast::{
+    NONE,
+    ast::{
+        BindingPatternKind, CatchClause, Declaration, ExportDefaultDeclarationKind, Expression,
+        ForStatementLeft, ImportOrExportKind, ModuleExportName, Program, SimpleAssignmentTarget,
+        Statement, VariableDeclarationKind,
+    },
+};
+use oxc_ecmascript::BoundNames;
+use oxc_semantic::ScopeId;
+use oxc_semantic::{ScopeFlags, SymbolFlags};
+use oxc_span::{Atom, SPAN};
+use oxc_traverse::{BoundIdentifier, Traverse};
```
Style nits:

For brevity, please combine `use` statements, so there's only 1 for each crate.

For `oxc_ast`, just blanket import all the AST types with `use oxc_ast::ast::*;`.

Throughout the visitors, please just `node: &mut ForOfStatement<'a>` instead of `node: &mut oxc_ast::ast::ForOfStatement<'a>`.

Ditto for `TraverseCtx` - import it at top of file, to avoid repetition of `&mut oxc_traverse::TraverseCtx<'a>`.

Ditto `Vec`. Please import at the top with `use oxc_allocator::Vec as ArenaVec;`. That's our naming convention in transformer to make it clear what's an arena `Vec` and what's a `std::vec::Vec`.

camc314 replied:
makes sense. should be fixed now - thanks

overlookmotel commented on crates/oxc_transformer/src/proposals/explicit_resource_management.rs:
```diff
@@ -0,0 +1,768 @@
+        if let Some(replacement) =
+            self.transform_statements(&mut node.body, None, scope_id, scope_id, ctx)
+        {
+            node.body = ctx.ast.vec_from_array([replacement]);
```
```suggestion
            node.body = ctx.ast.vec1(replacement);
```

overlookmotel commented on crates/oxc_transformer/src/proposals/explicit_resource_management.rs:
```diff
@@ -0,0 +1,768 @@
+            ctx.current_hoist_scope_id(),
+            ctx,
+        ) {
+            node.statements = ctx.ast.vec_from_array([replacement]);
```
```suggestion
            node.statements = ctx.ast.vec1(replacement);
```

overlookmotel reviewed (commented):
Thanks very much for tackling this. And bravo for getting all the tests to pass, including all the semantic IDs (don't worry about that 1 weird one).

I've made comments below about some small things.

The bigger thing is the performance hit. My assumption is it's due to looping over statements. Usually we try to avoid that by setting some state in `enter_statements`, then let the traversal run through the children with smaller visitors which update that state as they go, and then you check the state in `exit_statements` and act on it if there's work to be done. This avoids visiting every statement twice.

However, I'm not familiar with this transform, and found it hard to follow what the code here is doing, so can't judge how amenable it'd be to that approach. So I suggest we merge this, and then work on the performance in later PRs.

What would help a lot is if you'd be able to add more comments, so the logic is easier to follow. The `AstBuilder` calls are so verbose (bad API, sorry!) that it helps a lot to have a comment before a bunch of `ctx.ast.blah_blah` calls saying what the code it produces. e.g. add a comment before this:

```rs
// `var var_id = <expr>;`
```

https://github.com/oxc-project/oxc/blob/69a7a415f4ddaf080d9131454ee1a31bd98a31e5/crates/oxc_transformer/src/proposals/explicit_resource_management.rs#L270-L289

Also methods should ideally have a high-level comment with a before vs after example (exactly like you did on `enter_for_of_statement`).

Once the logic of the transform is clearer, it'll be easier to see what perf optimization is possible.

Sorry for the deluge of feedback. I hope it's helpful, rather than annoying!

camc314 commented on crates/oxc_transformer/examples/transformer.rs:
```diff
@@ -45,6 +45,7 @@ fn main() {
     let ret = SemanticBuilder::new()
         // Estimate transformer will triple scopes, symbols, references
         .with_excess_capacity(2.0)
+        .with_scope_tree_child_ids(true)
```
i need to think more here, i'm using it https://github.com/oxc-project/oxc/blob/fc2520a33a5be347a55ed1d48aec397114e099c1/crates/oxc_transformer/src/proposals/explicit_resource_management.rs#L414, but there might be a better way.

will look properly tommorow

camc314 replied:
yeah so if you run the example and it transforms a `using` declaration, then it'll panic because the child ids don't exist.

let me look and see if it's possible to do it another way

camc314 replied:
what's this trying to achieve?
1. we've got a vec of `Statements`
2. each of these statements currently has a parent scope of `x`
3. we are moving all of these statements into a child block
4. so we need to update the scope of all of the statements

e.g.
```
function foo () {}
```
becomes
```
try {
    function foo() {}
} ...
```

The current approach:

https://github.com/oxc-project/oxc/blob/c1b0ed3232ee38f41abc058a9b92e3376e4456fc/crates/oxc_transformer/src/proposals/explicit_resource_management.rs#L678-L690

gets all child scopes, moves them to the new scope id (ignoring the scope we just created.

would it make sense to introduce a new api in oxc_traverse to achieve this?

e.g. a similar (but new) version of. something like `insert_scope_above_statements`

https://github.com/oxc-project/oxc/blob/a5cde10ecc9cf5125cea7e44081fde334215e67b/crates/oxc_traverse/src/context/scoping.rs#L104-L115

^^ this is a bit of a brain dump.

Probably a better solution is to use `change_parent_id` on the scope we passed in, and replace it.
needs more thought.

overlookmotel replied:
> yeah so if you run the example and it transforms a using declaration, then it'll panic because the child ids don't exist.

Ah of course. Now I understand!

I may be wrong, but I don't think using child scopes will work in all cases. Because of e.g.:

```js
function outer(x = function inParams() {}) {
    using u = whatever( function inBody() {} );
}
```

Functions only have 1 scope, covering both the function body and its params. So in this case, `inParams`'s scope is a child of `outer`. `inBody` needs to have its parent scope changed, but `inParams` should not. So, if I'm right, we'll need another mechanism to determine which scopes to re-parent.

But let's leave as is for now, and I'll try to devise a failing test in a follow-up PR.

Also, fixing #9666 is going to throw all scopes-related stuff into disarray, so probably best we tackle this after that's done.

camc314 commented:
Thank you for the detailed review! I appreciate it.

For the perf hit, i had this thought. Would it be something like:
1. enter var decl.
2. if using, add address of parent (e.g. block stmt) to a hash map
3. when leaving a block, check if the address is in the hash map, if so, do some work. else skip.

If that makes sense, perhapse we merge this once the other feedback has been resolved and fixed in a foillowup?

I've actioned most of your comments just now. the more complex ones i need to think about more before actioning, but i'll try to do tomorrow/friday.

Boshen commented:
This is a great help for future maintenance: https://github.com/oxc-project/oxc/blob/main/crates/oxc_transformer/README.md#style-guide-for-implementing-transforms

overlookmotel commented:
Just to say, sorry I've been slow to come back. I'm abroad at the moment. I've been looking through it today, but need to look again with fresh eyes tomorrow. Will come back to you soon as I can.

overlookmotel reviewed (approved):
Unfortunately once #9666 is fixed, it'll require lots of changes to scopes. And we should be able to improve perf. But let's merge now and then iterate on it.

I'll follow up with a few PRs for minor stuff which is faster to do than to explain.

Nice one! Not an easy transform...

overlookmotel commented:
### Merge activity

* **Mar 11, 12:57 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel commented:
I made some tweaks to take into account recent changes to `Semantic`, so it compiles on latest main, but haven't made any substantive changes.

camc314 committed:
feat(transformer): transform explicit resource management (#9310)

~~wanted to have a shot at implementing this, this is rough draft.~~

~~Dunqing or overlookmotel feel free to take over or close out this PR. not sure how much time i'll have to finish it off~~

@Dunqing @overlookmotel this should be ready when you guys have time 🙏

closes #9168

overlookmotel commented:
> For the perf hit, i had this thought. Would it be something like:
> 
> 1. enter var decl.
> 2. if using, add address of parent (e.g. block stmt) to a hash map
> 3. when leaving a block, check if the address is in the hash map, if so, do some work. else skip.

Yes, pretty much exactly that. Except I'd suggest using a stack instead of a hashmap. You'd push to the stack when entering a block, and pop from it when exiting. In between, `VariableDeclaration` visitor would update the entry on top of the stack if it's a `using` statement.

Same principle, but a stack is usually cheaper than a hash map.

`oxc_data_structures` crate has some stack types which are optimized for this kind of thing. In this case, probably `SparseStack<BoundIdentifier>` would be a good choice (where the `BoundIdentifier` represents the binding for the `_usingCtx` var).

`SparseStack<BoundIdentifier>` behaves the same as a `Vec<Option<BoundIdentifier>>`, but optimized for when most entries are `None` - which they probably are, as `using` is a bit of niche language feature.

camc314 commented:
yeah best to try it out, and see what happens
